server/src/tracker.rs

```diff
@@ -0,0 +1,271 @@
+    inner: future::Abortable<F>,
+
+    reg: TrackerRegistry<T>,
+    id: TrackerId,
+}
+
+impl<F: Future, T> Future for TrackedFuture<F, T> {
+    type Output = Result<F::Output, future::Aborted>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.project().inner.poll(cx)
+    }
+}
+
+#[pinned_drop]
+impl<F: Future, T> PinnedDrop for TrackedFuture<F, T> {
+    fn drop(self: Pin<&mut Self>) {
+        // Avoid double panic by "leaking" Tracker
+        if !std::thread::panicking() {
+            let this = self.project();
+            this.reg.untrack(this.id)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tokio::sync::oneshot;
+
+    #[tokio::test]
+    async fn test_lifecycle() {
+        let (sender, receive) = oneshot::channel();
+        let reg = TrackerRegistry::new();
+
+        let task = tokio::spawn(receive.track(&reg, ()));
+
+        assert_eq!(reg.tracked().len(), 1);
+
+        sender.send(()).unwrap();
+        task.await.unwrap().unwrap().unwrap();
+
+        assert_eq!(reg.tracked().len(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_interleaved() {
+        let (sender1, receive1) = oneshot::channel();
+        let (sender2, receive2) = oneshot::channel();
+        let reg = TrackerRegistry::new();
+
+        let task1 = tokio::spawn(receive1.track(&reg, 1));
+        let task2 = tokio::spawn(receive2.track(&reg, 2));
+
+        let mut tracked: Vec<_> = reg.tracked().iter().map(|x| x.1).collect();
+        tracked.sort_unstable();
+        assert_eq!(tracked, vec![1, 2]);
+
+        sender2.send(()).unwrap();
+        task2.await.unwrap().unwrap().unwrap();
+
+        let tracked: Vec<_> = reg.tracked().iter().map(|x| x.1).collect();
+        assert_eq!(tracked, vec![1]);
+
+        sender1.send(42).unwrap();
+        let ret = task1.await.unwrap().unwrap().unwrap();
+
+        assert_eq!(ret, 42);
+        assert_eq!(reg.tracked().len(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_drop() {
+        let reg = TrackerRegistry::new();
+
+        {
+            let f = futures::future::pending::<()>().track(&reg, ());
+
+            assert_eq!(reg.tracked().len(), 1);
+
+            std::mem::drop(f);
+        }
+
+        assert_eq!(reg.tracked().len(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_terminate() {
+        let reg = TrackerRegistry::new();
+
+        let task = tokio::spawn(futures::future::pending::<()>().track(&reg, ()));
+
+        let tracked = reg.tracked();
+        assert_eq!(tracked.len(), 1);
+
+        reg.terminate(tracked[0].0);
+        let result = task.await.unwrap();
+
+        assert!(result.is_err());
+        assert_eq!(reg.tracked().len(), 0);
+    }
+
+    #[test]
+    fn test_panic() {
+        let reg = TrackerRegistry::new();
+
+        let result = std::panic::catch_unwind(|| {
+            #[allow(unused_variables)]
+            let f = futures::future::pending::<()>().track(&reg, ());
```

Comment:
Can this be:

```suggestion
            let _f = futures::future::pending::<()>().track(&reg, ());
```

instead?